### PR TITLE
fix(): rewrite nullable schemas for OpenAPI 3.1 documents

### PR DESCRIPTION
**Pull request by official-burak**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/nestjs/nest/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Other... Please describe:

## What is the current behavior?

\`DocumentBuilder.setOpenAPIVersion('3.1.0')\` sets \`openapi: 3.1.0\`, but schema properties still emit the OpenAPI 3.0 \`nullable\` keyword:

\`\`\`ts
@ApiProperty({ type: String, nullable: true })
// { \"type\": \"string\", \"nullable\": true }
\`\`\`

OpenAPI 3.1 uses JSON Schema 2020-12, which removed \`nullable\`. Strict 3.1 consumers therefore treat the property as a non-nullable string.

Issue Number: #4063

## What is the new behavior?

When the document version is 3.1 or later, \`SwaggerModule.createDocument\` rewrites:

- \`{ type: \"string\", nullable: true }\` to \`{ type: [\"string\", \"null\"] }\`
- nested property schemas the same way
- \`{ nullable: true, allOf: [{ \$ref }] }\` to \`{ oneOf: [allOf, { type: \"null\" }] }\`

3.0 documents are unchanged.

## Does this PR introduce a breaking change?
- [ ] Yes
- [x] No

Only 3.1+ documents change, and only to match the 3.1 schema language.

## Other information

Unit tests cover scalar unions, type arrays, nested properties, and nullable \`$ref\`/\`allOf\` schemas.

### CODE REVIEW

File: lib/utils/rewrite-nullable-for-oas31.util.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+  if (obj.nullable !== true) {
+    return;
+  }
+  delete obj.nullable;
+
+  if (typeof obj.type === 'string') {
+    obj.type = [obj.type, 'null'];
+    return;
+  }
```
Comment (simonboeuf-kpler) [inline lib/utils/rewrite-nullable-for-oas31.util.ts:29]: We do not yet cover a typed `$ref`/`allOf` composite (the shape the factory emits for `@ApiProperty({ type: Dto, nullable: true })`). It becomes `{ type: ['object','null'], allOf: [...] }`, which can never accept null: the type allows null, but the allOf still forbids it.

The fix guards both type branches so composites fall through to the `oneOf` wrap you already have. The same block also corrects `type: 'null'`, which currently produces the invalid `['null','null']`. It replaces the two type checks (lines 26-35):

```ts
  const isComposite =
    '$ref' in obj || 'allOf' in obj || 'anyOf' in obj || 'oneOf' in obj;

  if (typeof obj.type === 'string' && !isComposite) {
    obj.type = obj.type === 'null' ? 'null' : [obj.type, 'null'];
    return;
  }
  if (Array.isArray(obj.type) && !isComposite) {
    if (!obj.type.includes('null')) {
      obj.type = [...obj.type, 'null'];
    }
    return;
  }
```

Failing tests:

```typescript
it('wraps a typed nullable $ref/allOf composite in oneOf', () => {
  const schema: any = { type: 'object', nullable: true, allOf: [{ $ref: '#/components/schemas/Profile' }] };
  rewriteNullableForOas31(schema);
  expect(schema).toEqual({ oneOf: [{ type: 'object', allOf: [{ $ref: '#/components/schemas/Profile' }] }, { type: 'null' }] });
});

it('does not produce a ["null","null"] union', () => {
  const schema: any = { type: 'null', nullable: true };
  rewriteNullableForOas31(schema);
  expect(schema).toEqual({ type: 'null' });
});
```

File: lib/utils/rewrite-nullable-for-oas31.util.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+  const obj = value as Record<string, unknown>;
+  for (const nested of Object.values(obj)) {
+    rewriteNullableForOas31(nested);
+  }
+
+  if (obj.nullable !== true) {
+    return;
+  }
+  delete obj.nullable;
+
```
Comment (simonboeuf-kpler) [inline lib/utils/rewrite-nullable-for-oas31.util.ts:25]: We should handle both `enum` and `const`. Failing tests example: 

```typescript
it('adds null to enum values', () => {
  const schema: any = { type: 'string', enum: ['a', 'b'], nullable: true };
  rewriteNullableForOas31(schema);
  expect(schema).toEqual({ type: ['string', 'null'], enum: ['a', 'b', null] });
});
it('keeps a nullable const satisfiable', () => {
  const schema: any = { type: 'string', const: 'x', nullable: true };
  rewriteNullableForOas31(schema);
  expect(schema).toEqual({ type: ['string', 'null'], enum: ['x', null] });
});
```



```suggestion
 if (Array.isArray(obj.enum)) {
    if (!obj.enum.includes(null)) {
      obj.enum = [...obj.enum, null];
    }
  } else if ('const' in obj) {
    obj.enum = [obj.const, null];
    delete obj.const;
  }
```

File: lib/utils/rewrite-nullable-for-oas31.util.ts [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+  if (typeof obj.type === 'string') {
+    obj.type = [obj.type, 'null'];
```
Comment (simonboeuf-kpler) [inline lib/utils/rewrite-nullable-for-oas31.util.ts]: For type: `'null'` this produces `['null','null']`, which is not valid 

Failing test example:

```
it('does not produce a ["null","null"] union', () => {
  const schema: any = { type: 'null', nullable: true };
  rewriteNullableForOas31(schema);
  expect(schema).toEqual({ type: 'null' });
});
```


```suggestion
    obj.type = obj.type === 'null' ? 'null' : [obj.type, 'null'];
```

File: lib/utils/rewrite-nullable-for-oas31.util.ts [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+  }
+
+  const obj = value as Record<string, unknown>;
+  for (const nested of Object.values(obj)) {
+    rewriteNullableForOas31(nested);
+  }
+
+  if (obj.nullable !== true) {
+    return;
+  }
```
Comment (simonboeuf-kpler) [inline lib/utils/rewrite-nullable-for-oas31.util.ts]: This leaves `nullable: false` in the document. Acceptable but we can easily drop it here. Failing test: 

```typescript
it('drops a redundant nullable: false', () => {
  const schema: any = { type: 'string', nullable: false };
  rewriteNullableForOas31(schema);
  expect(schema).toEqual({ type: 'string' });
});
```


```suggestion
  if (obj.nullable !== true) {
    delete obj.nullable;
    return;
  }
```

File: lib/utils/rewrite-nullable-for-oas31.util.ts [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    obj.type = [obj.type, 'null'];
+    return;
+  }
+  if (Array.isArray(obj.type)) {
+    if (!obj.type.includes('null')) {
+      obj.type = [...obj.type, 'null'];
+    }
+    return;
+  }
+
```
Comment (simonboeuf-kpler) [inline lib/utils/rewrite-nullable-for-oas31.util.ts:36]: When a `oneOf/anyOf` already has a null branch, the wrap below nests it a second time. This guard skips the redundant wrap. It needs one small helper at module scope.

Failing test example: 

```typescript
it('drops a redundant nullable when the union already admits null', () => {
  const schema: any = { nullable: true, oneOf: [{ type: 'string' }, { type: 'null' }] };
  rewriteNullableForOas31(schema);
  expect(schema).toEqual({ oneOf: [{ type: 'string' }, { type: 'null' }] });
});
```
 
Suggested fix

```
  const branches = [obj.oneOf, obj.anyOf].find(Array.isArray) as unknown[] | undefined;
  if (branches?.some(admitsNull)) {
    return;
  }
  
// helper function, probably at module scope
function admitsNull(schema: unknown): boolean {
  if (!schema || typeof schema !== 'object') {
    return false;
  }
  const s = schema as Record<string, unknown>;
  return s.type === 'null' || (Array.isArray(s.type) && s.type.includes('null'));
}
```

File: test/utils/rewrite-nullable-for-oas31.util.spec.ts [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    });
+  });
+
+  it('leaves 3.0-compatible schemas without nullable unchanged', () => {
+    const schema: any = { type: 'string' };
+
+    rewriteNullableForOas31(schema);
+
+    expect(schema).toEqual({ type: 'string' });
+  });
```
Comment (simonboeuf-kpler) [inline test/utils/rewrite-nullable-for-oas31.util.spec.ts:63]: This recursion goes into every value, including `example`, `examples`, `default`, `const` payloads, and `x-*` extensions. A correct fix needs position-aware traversal, which is a larger change, just flagging this as a failing test here, let me know if you would like a fix suggestion

```suggestion
  });
  
  it('does not rewrite a schema-shaped example payload', () => {
  const schema: any = { type: 'object', properties: { name: { type: 'string' } }, example: { note: { type: 'string', nullable: true } } };
  rewriteNullableForOas31(schema);
  expect(schema.example).toEqual({ note: { type: 'string', nullable: true } });
});
```

File: lib/utils/rewrite-nullable-for-oas31.util.ts [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+  const obj = value as Record<string, unknown>;
+  for (const nested of Object.values(obj)) {
+    rewriteNullableForOas31(nested);
+  }
+
+  if (obj.nullable !== true) {
+    return;
+  }
+  delete obj.nullable;
```
Comment (simonboeuf-kpler) [inline lib/utils/rewrite-nullable-for-oas31.util.ts]: This leaves `nullable: false` in the document. Acceptable but we can easily drop it here. Failing test: 

```typescript
it('drops a redundant nullable: false', () => {
  const schema: any = { type: 'string', nullable: false };
  rewriteNullableForOas31(schema);
  expect(schema).toEqual({ type: 'string' });
});
```


```suggestion
  if (obj.nullable !== true) {
    delete obj.nullable;
    return;
  }
```

File: lib/utils/rewrite-nullable-for-oas31.util.ts (hunk 4 above)
Comment (simonboeuf-kpler) [inline lib/utils/rewrite-nullable-for-oas31.util.ts:23]: This leaves `nullable: false` in the document. Acceptable but we can easily drop it here. Failing test: 

```typescript
it('drops a redundant nullable: false', () => {
  const schema: any = { type: 'string', nullable: false };
  rewriteNullableForOas31(schema);
  expect(schema).toEqual({ type: 'string' });
});
```


```suggestion
  if (obj.nullable !== true) {
    delete obj.nullable;
    return;
  }
```

**Review by simonboeuf-kpler — COMMENTED** [review]
Thanks for the PR! 

We run our own 3.1 nullable normalization so we have hit several edge cases in real specs. I added an inline comment for each one, and every comment has a failing test and a suggested fix.

The last comment, on the spec file, is a heads-up only. A correct fix needs position-aware traversal, which is a larger change. I would be glad to help with that if it is useful.

**Comment by micalevisk** [comment]
Thanks for taking this on, and thanks @simonboeuf-kpler for the detailed inline review.

Overall: **right approach, under-built.** A post-generation document walker is the correct shape for this fix — it catches everything, including the `nullable: true` that the CLI plugin injects, which a factory-level fix would miss. But the current implementation has correctness bugs on the exact shapes this library emits, plus one design flaw that can't be patched incrementally.

## On the existing inline comments

All five are legitimate. Ranked by how much they should block:

| # | Comment | Severity |
|---|---|---|
| 1 | Typed `$ref`/`allOf` composite → `{type:['object','null'], allOf:[…]}` | **Blocking** |
| 2 | Blind recursion into `example` / `default` / `x-*` | **Blocking (design)** |
| 3 | `enum` / `const` not extended with `null` | High |
| 4 | Double-wrap when `oneOf` already has a null branch | Medium |
| 5 | `nullable: false` left in the document | Low |

**`#1` is the headline bug, and it's provable — not hypothetical.** `lib/services/schema-object-factory.ts#L649-L656` emits exactly that shape for `@ApiProperty({ type: SomeDto, nullable: true })`:

```ts
...validMetadataObject,
...(validMetadataObject['nullable'] ? { type: 'object' } : {}),
allOf: [{ $ref }]
```

The `typeof obj.type === 'string'` branch fires first, producing `{ type: ['object','null'], allOf: [{ $ref }] }`. That is *unsatisfiable for null*: the type union admits `null`, but the `allOf` branch still requires the referenced object. So the single most common nullable-DTO case ends up broken in a way that's arguably worse than the status quo. The `isComposite` guard suggested above is the right fix.

**`#2` needs a structural fix, not a guard.** `rewriteNullableForOas31` recurses into every object value with no notion of *where* it is in the document. Any `example`, `examples`, `default`, or `x-*` extension whose payload happens to contain a `nullable: true` key gets rewritten as if it were a schema — a user documenting their own OpenAPI-ish config object would get their example silently corrupted. The fix is a position-aware walk: `paths` → path items → operations → parameters / requestBody / responses → content → `schema`, plus `components.*` and `webhooks`, applying the conversion only at genuine schema positions and recursing into `properties`, `items`, `additionalProperties`, `prefixItems`, `not`, and the combinators. That's ~150 lines instead of 42, but it's the only version that's actually safe.

## A few things not yet raised

**Annotation keywords get swallowed by the `oneOf` wrap.** The catch-all branch moves *every* sibling key inside the union, so `{ nullable: true, description: 'Owner', allOf: [{ $ref }] }` becomes `{ oneOf: [{ description: 'Owner', allOf: […] }, { type: 'null' }] }`. `description`, `title`, `deprecated`, and `readOnly` still apply at the union level and should stay outside it — otherwise tooling that reads the property description loses it.

**The wrap fires on schemas that don't need it.** `{ nullable: true, description: 'x' }` — no `type`, no `$ref` — becomes `{ oneOf: [{ description: 'x' }, { type: 'null' }] }`. A typeless schema already admits `null`; the correct action there is to drop `nullable` and return.

**In-place mutation of possibly shared objects.** The walker mutates the document deeply, but `documentForVersion` is only a *shallow* spread of `baseDocument`, and the factory builds schemas with `omit(...)`, which is also shallow — so nested objects such as `metadata.items` and `metadata.enumSchema` may still be the same references held in the decorator metadata. An app that calls `createDocument` twice (a 3.1 doc and a 3.0 doc, or one document per API version — not exotic) risks the 3.1 rewrite leaking into the 3.0 document. Worth a regression test: build a 3.1 document, then a 3.0 document from the same controllers, and assert the 3.0 one still has `nullable: true`.

**Nothing tests the version gate.** All five tests call the util directly. Nothing asserts that `createDocument` applies it for `3.1.0` and skips it for `3.0.0` — which is the actual contract of this fix. @simonboeuf-kpler offered e2e tests in the issue; an e2e case that generates both a 3.0 and a 3.1 document from the same fixture app would cover both this and the shared-mutation point above.

## Out of scope here, but adjacent

The reverse direction is still broken and this PR shouldn't try to fix it — flagging it so it isn't assumed solved once this merges. Since #3897 the response object factory emits `oneOf: [<schema>, { type: 'null' }]` **unconditionally**, so **3.0 documents get 3.1 syntax** in inline responses. That means the "one document mixes both dialects" complaint in #4063 is only half-addressed by this PR: it fixes 3.1 documents carrying 3.0 spelling, but 3.0 documents carrying 3.1 spelling remains. The proper fix is to thread the declared OpenAPI version down through `SwaggerScanner` → `SwaggerExplorer` → `ResponseObjectFactory` → `ResponseObjectMapper` so the response path can pick the correct spelling per version, defaulting to `3.0.0` for backward compatibility. Separate PR.

## What I'd like to see before merge

- #1 and #3 fixed (both correctness)
- #4 and #5 taken as suggested (cheap)
- #2 restructured into a position-aware walker
- an e2e test generating both a 3.0 and a 3.1 document from the same fixture app

One offer to save you the work on `#2`: I have a working position-aware walker locally that already handles the composite case, `enum` null-extension, annotation preservation, and unwinding the synthetic `{ type: 'object', allOf: [{ $ref }] }` wrapper. Happy to hand it over so you can fold it into this PR rather than rebuilding it — just say the word.

**Comment by kamilmysliwiec** [comment]
ref https://github.com/nestjs/swagger/pull/4070
